File: modules/agent-go/internal/repositories/account_repository.go
package repositories

import (
	"github.com/pina-colada-co/agent-go/internal/models"
	"gorm.io/gorm"
)

type AccountRepository struct {
	db *gorm.DB
}

func NewAccountRepository(db *gorm.DB) *AccountRepository {
	return &AccountRepository{db: db}
}

// AccountSearchResult represents an account search result
type AccountSearchResult struct {
	ID        int64
	AccountID int64
	Name      string
	Type      string
}

func (r *AccountRepository) Search(query string, tenantID *int64, limit int) ([]AccountSearchResult, error) {
	var accounts []models.Account
	q := r.db.Preload("Organizations").Preload("Individuals").
		Where("LOWER(name) LIKE LOWER(?)", "%"+query+"%")
	if tenantID != nil {
		q = q.Where("tenant_id = ?", *tenantID)
	}
	if err := q.Order("name").Limit(limit).Find(&accounts).Error; err != nil {
		return nil, err
	}

	results := make([]AccountSearchResult, len(accounts))
	for i, a := range accounts {
		accountType, entityID := getAccountTypeAndEntityID(a)
		results[i] = AccountSearchResult{
			ID:        entityID,
			AccountID: a.ID,
			Name:      a.Name,
			Type:      accountType,
		}
	}
	return results, nil
}

func getAccountTypeAndEntityID(account models.Account) (string, int64) {
	if len(account.Organizations) > 0 {
		return "organization", account.Organizations[0].ID
	}
	if len(account.Individuals) > 0 {
		return "individual", account.Individuals[0].ID
	}
	return "unknown", account.ID
}

// CreateRelationship creates a relationship between two accounts
func (r *AccountRepository) CreateRelationship(fromAccountID, toAccountID, userID int64, relationshipType, notes *string) (*models.AccountRelationship, error) {
	rel := &models.AccountRelationship{
		FromAccountID:    fromAccountID,
		ToAccountID:      toAccountID,
		RelationshipType: relationshipType,
		Notes:            notes,
		CreatedBy:        userID,
		UpdatedBy:        userID,
	}
	if err := r.db.Create(rel).Error; err != nil {
		return nil, err
	}
	return rel, nil
}

// DeleteRelationship deletes a relationship
func (r *AccountRepository) DeleteRelationship(fromAccountID, relationshipID int64) error {
	return r.db.Where("id = ? AND (from_account_id = ? OR to_account_id = ?)", relationshipID, fromAccountID, fromAccountID).
		Delete(&models.AccountRelationship{}).Error
}

File: modules/agent/src/repositories/account_repository.py
"""Repository layer for account data access."""

import logging
from typing import List, Optional

from sqlalchemy import select
from sqlalchemy.orm import selectinload

from lib.db import async_get_session
from models.Account import Account
from models.AccountRelationship import AccountRelationship

logger = logging.getLogger(__name__)


async def search_accounts(query: str, tenant_id: Optional[int]) -> List[Account]:
    """Search accounts by name."""
    async with async_get_session() as session:
        search_pattern = f"%{query}%"
        stmt = (
            select(Account)
            .options(selectinload(Account.organizations), selectinload(Account.individuals))
            .where(Account.name.ilike(search_pattern))
            .order_by(Account.name)
            .limit(20)
        )
        if tenant_id:
            stmt = stmt.where(Account.tenant_id == tenant_id)
        result = await session.execute(stmt)
        return list(result.scalars().all())


async def find_account_relationship(from_account_id: int, to_account_id: int) -> Optional[AccountRelationship]:
    """Find existing relationship between two accounts."""
    async with async_get_session() as session:
        stmt = select(AccountRelationship).where(
            AccountRelationship.from_account_id == from_account_id,
            AccountRelationship.to_account_id == to_account_id,
        )
        result = await session.execute(stmt)
        return result.scalar_one_or_none()


async def create_account_relationship(
    from_account_id: int,
    to_account_id: int,
    user_id: int,
    relationship_type: Optional[str] = None,
    notes: Optional[str] = None,
) -> AccountRelationship:
    """Create a new account relationship."""
    async with async_get_session() as session:
        relationship = AccountRelationship(
            from_account_id=from_account_id,
            to_account_id=to_account_id,
            relationship_type=relationship_type,
            notes=notes,
            created_by=user_id,
            updated_by=user_id,
        )
        session.add(relationship)
        await session.commit()
        await session.refresh(relationship)
        return relationship


async def find_account_relationship_by_id(
    relationship_id: int, account_id: int
) -> Optional[AccountRelationship]:
    """Find relationship by ID where account is either from or to."""
    async with async_get_session() as session:
        stmt = select(AccountRelationship).where(
            AccountRelationship.id == relationship_id,
            (AccountRelationship.from_account_id == account_id) |
            (AccountRelationship.to_account_id == account_id),
        )
        result = await session.execute(stmt)
        return result.scalar_one_or_none()


async def delete_account_relationship(relationship: AccountRelationship) -> None:
    """Delete an account relationship."""
    async with async_get_session() as session:
        # Re-fetch in this session
        rel = await session.get(AccountRelationship, relationship.id)
        if rel:
            await session.delete(rel)
            await session.commit()


async def get_account_relationships(account_id: int) -> List[AccountRelationship]:
    """Get all relationships for an account (both outgoing and incoming)."""
    async with async_get_session() as session:
        stmt = (
            select(AccountRelationship)
            .options(
                selectinload(AccountRelationship.from_account).selectinload(Account.organizations),
                selectinload(AccountRelationship.from_account).selectinload(Account.individuals),
                selectinload(AccountRelationship.to_account).selectinload(Account.organizations),
                selectinload(AccountRelationship.to_account).selectinload(Account.individuals),
            )
            .where(
                (AccountRelationship.from_account_id == account_id) |
                (AccountRelationship.to_account_id == account_id)
            )
        )
        result = await session.execute(stmt)
        return list(result.scalars().all())
